Automatically retry with correct content-type

trevor-vaughan opened the pull request:
The podman socket will return a note about the correct content-type that
should be used for a given type of message if it is incorrect.

This is a workaround to automatically correct the content-type and
resubmit the call.

A full correction would be to update all of the different connection
entries but this appears to work correctly in the short term.

Fixes #568

trevor-vaughan commented:
The errors in Travis appear to be related to the busybox image and not the code changes as far as I can tell.

trevor-vaughan commented:
Ok, so the Travis tests are failing (for whatever reason), but everything [passes in GitHub Actions](https://github.com/trevor-vaughan/docker-api/runs/1802694288?check_suite_focus=true) which I added to this PR to see if it would help.

trevor-vaughan commented:
Sorry folks, the different ruby versions are being troublesome. Will update again once I have the GitHub Actions working properly.

trevor-vaughan commented:
Ok, this is now [correctly passing in GitHub Actions](https://github.com/trevor-vaughan/docker-api/actions/runs/527867238) and should be ready for review.

trevor-vaughan commented:
@phaelyn @tlunter Hi folks. Sorry for the bother, but I just wanted to check and see if you're still maintaining the gem and/or open to contributions. The PR list seems pretty extensive and hasn't really shifted much from what I can tell.

tlunter commented:
@trevor-vaughan can i ask why you decided to add github actions to the PR? we are already using travis for the testing of the repo.

trevor-vaughan commented:
@tlunter Well, the main reason is that I couldn't get them to work in Travis. I (and everyone else at this point) can guarantee that they will be run in GitHub. Once I did the work, I figured I should go ahead and submit it.

Also, I'm not sure if you were hit by this (yet) but Travis' support of FOSS projects is now severely lacking with no apparent exit plan.

If you don't like it, I can certainly back it out, but it works quite well (particularly for contributors).

tlunter commented on lib/docker/image.rb:
```diff
@@ -227,7 +227,16 @@ def search(query = {}, connection = Docker.connection, creds = nil)
+      require 'open-uri'
+
+      # This differs after Ruby 2.4
+      if URI.public_methods.include?(:open)
```
Is this just `#respond_to?`?

trevor-vaughan replied:
It's not actually. It `respond_to?` returns true in both pre 2.4 and post 2.4 so we need to check and see if it's an actual public method.

tlunter commented on lib/docker/image.rb:
```diff
@@ -227,7 +227,16 @@ def search(query = {}, connection = Docker.connection, creds = nil)
     # argument may either be a File or URI.
     def import(imp, opts = {}, conn = Docker.connection)
-      open(imp) do |io|
+      require 'open-uri'
```
I don't frequently open IO objects beside local files, what's the benefit here vs the prior `open` method?

trevor-vaughan replied:
Seamless Ruby 2.7 compatibility.

tlunter commented on lib/docker/connection.rb:
```diff
@@ -35,21 +35,51 @@ def resource
+          #   * 'Content-Type: application/json is not supported. Should be "application/x-tar"'
+          matches = response_cause.delete('"\'').scan(%r{(application/\S+)})
+          unless matches.count < 2
+            request[:headers]['Content-Type'] = matches.last.first
```
This is pretty spooky to me. Can we log something so that folks can clearly open an issue or something when this is hit? Also, is there a particular endpoint that you know needs to be fixed?

trevor-vaughan replied:
Not going to lie, it's 100% spooky. I'm fine with logging things if you have a preferred logging mechanism. I allowed updates from maintainers so feel free to hack away if that's easier and I can put it back through its paces in podman.

I took a swipe at the Docker API docs but multiple endpoints need to be updated and the downstream code doesn't actually support passing it in as an option.

As the comment mentions, this is basically a stop-gap until the real fix can be introduced later (which will probably involve a larger code overhaul if I'm reading things right).

tlunter replied:
I guess the limited ability for logging is from: https://github.com/swipely/docker-api/pull/569/files#diff-a8fbee6483c191f77ba681f51094f68035815545d435d51a195934263bd79850R86-R90

Doesn't really offer much since it's opt in...

trevor-vaughan replied:
Also, while this is spooky, it's "safe spooky" in that the worst case scenario is that the thing you were trying to do gets tried twice and doesn't work (and returns the usual error).

tlunter replied:
My fear mostly stems from the docker engine API being a bit harder to work with earlier on and potentially performing destructive behaviors, but not returning the right result. Agreed that it shouldn't actually do that here since it's failing at the API top-level.

trevor-vaughan replied:
Well, `docker` doesn't care. It seems to just ignore it and "do the right thing" for the operation. The `podman` API is strict and will fail.

tlunter commented:
I'd prefer we used the travis-ci stuff. There is a couple tests that run with the config there that's hidden. Agreed that we will probably need to migrate soon though, so even putting in another PR with the travis stuff removed and the actions stuff added would help.

trevor-vaughan commented:
@tlunter Would you also be OK with just leaving it and re-adding the travis stuff? You can change what you pay attention to in the project settings.

I literally couldn't get anything to run in Travis so I'd have to ping-pong it back and forth to figure out if it's still working.

trevor-vaughan commented:
@tlunter Do you prefer lots of little pushes with a merge/squash at the end or rebase with force push?

tlunter commented:
The former, typically with a standard merge, no squash. Would prefer squashing only when pushing to test something. Those can merged/rebased together.

trevor-vaughan commented:
OK, this obviously isn't making anyone's life hard but mine right now so I can fix things up this weekend if you're good with the general implementation.

I can split the github stuff into a separate commit as well if you'd like so it gets separated away from the actual code update.

tlunter commented:
That would be appreciated. We had done some work to fix the `busybox` issue in travis towards the end of last year, and now it's back... Going to see if I can dig up what our theory was there.

trevor-vaughan commented:
@tlunter Everything appears to be good to go now.

I went ahead and updated the rest of the tests to work with Podman and added podman tests to the GitHub Actions.

tlunter commented on lib/docker.rb:
```diff
@@ -119,6 +119,20 @@ def ping(connection = self.connection)
   end
 
+  # Determine if the server is podman or docker.
+  def podman?(connection = self.connection)
```
I wonder if we can instead cache this value on the connection. Right now we're going to perform this check every time we perform certain requests, which makes another API call to the version endpoint. A connection certainly isn't guaranteed to be fixed to a single docker endpoint, but that would mean swapping the daemon listening on the socket out from underneath the client, which would be a pretty odd thing to do...

trevor-vaughan replied:
@tlunter I'm certainly fine with that. Some people don't care for instance variables popping up everywhere but happy to do that if you'd like.

I agree that it would be super weird to swap it out in the middle of a program.

tlunter replied:
Yes please. Otherwise everything else looks good to me.

trevor-vaughan commented:
@tlunter Should be good to go. You might want to consider adding some of the other items in there as well. Things like `version()` and `info()` are good candidates.

trevor-vaughan commented:
@tlunter So, I seem to be getting some `cane` issues but I didn't touch anything that it's referring to. Should I just bump up the values? I didn't see any way to do line-by-line exclusions like you can in `rubocop`.

tlunter commented:
@trevor-vaughan I don't actually know where to look for the failures you're seeing since I only see travis on this PR, but yea, can just increase the limit for cane for now.

trevor-vaughan commented:
@tlunter Sorry about the flipping. Everything looks good now though. https://github.com/trevor-vaughan/docker-api/actions/runs/591109632

trevor-vaughan commented:
@tlunter Does this work for you now or did I miss something (pretty sure I got it all)?

tlunter commented on lib/docker.rb:
```diff
@@ -119,6 +119,29 @@ def ping(connection = self.connection)
 
+  # Determine if the server is podman or docker.
+  def podman?(connection = self.connection)
+    @attrs ||= {}
```
Sorry, I was thinking we could cache the `version` request coming from the connection. This is caching it on the top level module right now which can be shared across many connections.

tlunter replied:
https://github.com/trevor-vaughan/docker-api/compare/podman-compat...swipely:podman-compat-cache-version Like this.

tlunter commented on lib/docker.rb:
```diff
@@ -119,6 +119,29 @@ def ping(connection = self.connection)
+  end
+
+  # Determine if the session is rootless.
+  def rootless?(connection = self.connection)
```
Since this isn't used programmatically for other API calls, we can keep this one uncached.

trevor-vaughan replied:
I've ended up needing to use it quite a bit elsewhere so it makes sense to cache it to me.

trevor-vaughan commented:
@tlunter Well, in theory this is caching in the connection, not globally, across the board now. Unfortunately, I'm having a bit of fun with GitHub Actions not being able to decide which version of Ubuntu it wants to provide at the moment. The tests [worked fine in my branch](https://github.com/trevor-vaughan/docker-api/actions/runs/606568212).

tlunter commented on .github/workflows/unit_test.yml:
```diff
@@ -0,0 +1,82 @@
+          sudo add-apt-repository "deb [arch=amd64] https://download.docker.com/linux/ubuntu $(lsb_release -cs) stable"
+          sudo apt-get update -y
+          sudo apt-cache gencaches
+          sudo apt-get install -y docker-ce=$( apt-cache madison docker-ce | grep -e $DOCKER_VERSION | cut -f 2 -d '|' | head -1 | sed 's/\s//g' )
```
Maybe we can also grep here based on `lsb_release`? Are we expecting it to be `bionic` or `focal`? I assume `bionic` because of the `runs-on` for the `docker-rspec` section.

trevor-vaughan replied:
The repo setup on Line 45 targets it for whatever it's running. Installation outside of that  won't work.

`ubuntu-latest` seemed to give me varying results so I just gave up and queried whatever I had.

trevor-vaughan replied:
@tlunter I think this should be OK if the message from yesterday made sense :point_up:

tlunter replied:
Oh yes, definitely. Didn't see this was after the change that made the CI tests pass again. Thanks for resolving.

trevor-vaughan commented:
@tlunter Awesome! Is there anything else you need me to do before a merge (and hopefully release) is possible?

tlunter commented:
Released as v2.1.0. Thanks a lot @trevor-vaughan

trevor-vaughan commented:
@tlunter Thanks for putting up with all the noise and working it through with me!
